0000-product.md

```diff
@@ -0,0 +1,470 @@
+
+``` 
+“621dee” + “02” + “01” +“00000000000000000000000000000000000000000000” +
+14-character “numeric string” product_id + “00” // product_id == GTIN 
+```
+
+A full GS1 Product address using the example GTIN from https://www.gtin.info/
+would therefore be:
+
+``` 
+“621dee0201000000000000000000000000000000000000000000000001234560001200” 
+```
+
+## Transaction Payload and Execution
+
+### ProductPayload Transaction
+
+ProductPayload contains an action enum and the associated action payload.  This
+allows for the action payload to be dispatched to the appropriate logic.
+
+Only the defined actions are available and only one action payload should be
+defined in the ProductPayload.
+
+``` 
+message ProductPayload { 
+    enum Actions { 
+        UNSET_ACTION = 0; 
+        PRODUCT_CREATE = 1; 
+        PRODUCT_UPDATE = 2; 
+        PRODUCT_DELETE = 3; 
+    }
+
+    Action action = 1;
+
+    // Approximately when transaction was submitted, as a Unix UTC
+    // timestamp
+    uint64 timestamp = 2;
+
+    ProductCreateAction product_create = 3; 
+    ProductUpdateAction product_update = 4; 
+    ProductDeleteAction product_delete = 5; 
+} 
+```
+
+### ProductCreateAction
+
+ProductCreateAction adds a new product to state. The transaction should be
+submitted by an agent, which is identified by its signing key, acting on behalf
+of the organization that corresponds to the owner in the create transaction.
+(Organizations and agents are defined by the Pike smart contract.)
+
+``` 
+message ProductCreateAction { 
+    enum Product_Namespace { 
+        UNSET_NAMESPACE = 0;
+        GS1 = 1; 
+    }
+    // product_namespace and product_id are used in deriving the state address
+    Product_Namespace product_namespace = 1; 
+    string product_id = 2; 
+    string owner = 3; 
+    repeated PropertyValues properties = 4; 
+} 
+```
+
+Validation requirements:
+
+* If a product with product_id already exists the transaction is invalid.  
+* The signer of the transaction must be an agent in the Pike state and must
+belong to an organization in Pike state, otherwise the transaction is invalid.
+* The agent must have the permission can_create_product for the organization,
+otherwise the transaction is invalid.
+* If the product_namespace is GS1, the organization must contain a GS1 Company Prefix in its metadata (gs1_company_prefixes), and the prefix must match the company prefix in the product_id, which is a gtin if GS1, otherwise the transaction is invalid.  
+* The properties must be valid for the product_namespace. For example, if the product is GS1 product, its properties must only contain properties that are includedin the GS1 Schema. If it includes a property not in the GS1 Schema the transaction is invalid.  _The base GS1 schema will be defined in a future RFC._
+
+The product will be set in state.
+
+The inputs for ProductCreateAction must include:
+
+* Address of the Agent submitting the transaction Address of the Organization
```

Comment:
This bullet section is messed up. Several points are now running together and do not make sense.